### PR TITLE
Converts the captcha plugins to service providers

### PR DESCRIPTION
**Pull request by laoneo**
### Summary of Changes
Converts the captcha plugins to service providers.

### Testing Instructions
Setup a captcha plugin and edit an article in the front end.

### Actual result BEFORE applying this Pull Request
Captcha is shown.

### Expected result AFTER applying this Pull Request
Captcha is shown.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by dgrammatiko** [comment]
@laoneo is this an alternative to https://github.com/joomla/joomla-cms/pull/39657 or that one is still valid?

**Comment by laoneo** [comment]
Still valid... Can be used as followup of this one.

**Comment by MacJoom** [comment]
I have tested this item :white_check_mark: successfully on bb0f92940da2540b40f88db73eb47b773b9eaaee<br /><br />No changes detected - captcha works (PHP 8.1)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39729">issues.joomla.org/tracker/joomla-cms/39729</a>.</sub>

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on bb0f92940da2540b40f88db73eb47b773b9eaaee<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39729">issues.joomla.org/tracker/joomla-cms/39729</a>.</sub>

File: plugins/captcha/recaptcha/services/provider.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+     * @return  void
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function register(Container $container)
+    {
+        $container->set(
+            PluginInterface::class,
+            function (Container $container) {
+                $plugin     = new ReCaptcha(
```
Comment (richard67) [inline plugins/captcha/recaptcha/services/provider.php]:
```suggestion
                $plugin = new ReCaptcha(
```

File: plugins/captcha/recaptcha_invisible/services/provider.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+     * @return  void
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function register(Container $container)
+    {
+        $container->set(
+            PluginInterface::class,
+            function (Container $container) {
+                $plugin     = new InvisibleReCaptcha(
```
Comment (richard67) [inline plugins/captcha/recaptcha_invisible/services/provider.php]:
```suggestion
                $plugin = new InvisibleReCaptcha(
```

File: tests/Unit/Plugin/Captcha/InvisibleRecaptcha/Extension/InvisibleRecaptchaPluginTest.php [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+      * @testdox  can init the captcha
+      *
+      * @return  void
+      *
+      * @since   __DEPLOY_VERSION__
+      */
+    public function testInit()
+    {
+        $document   = new HtmlDocument();
+        $language   = $this->createStub(Language::class);
```
Comment (richard67) [inline tests/Unit/Plugin/Captcha/InvisibleRecaptcha/Extension/InvisibleRecaptchaPluginTest.php]:
```suggestion
        $document = new HtmlDocument();
        $language = $this->createStub(Language::class);
```

File: tests/Unit/Plugin/Captcha/InvisibleRecaptcha/Extension/InvisibleRecaptchaPluginTest.php [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    /**
+      * @testdox  can init the captcha with an empty public key
+      *
+      * @return  void
+      *
+      * @since   __DEPLOY_VERSION__
+      */
+    public function testInitEmptyPublicKey()
+    {
+        $language   = $this->createStub(Language::class);
```
Comment (richard67) [inline tests/Unit/Plugin/Captcha/InvisibleRecaptcha/Extension/InvisibleRecaptchaPluginTest.php]:
```suggestion
        $language = $this->createStub(Language::class);
```

File: tests/Unit/Plugin/Captcha/InvisibleRecaptcha/Extension/InvisibleRecaptchaPluginTest.php [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+    /**
+      * @testdox  can check successful answer
+      *
+      * @return  void
+      *
+      * @since   __DEPLOY_VERSION__
+      */
+    public function testResponse()
+    {
+        $language   = $this->createStub(Language::class);
```
Comment (richard67) [inline tests/Unit/Plugin/Captcha/InvisibleRecaptcha/Extension/InvisibleRecaptchaPluginTest.php]:
```suggestion
        $language = $this->createStub(Language::class);
```

File: tests/Unit/Plugin/Captcha/InvisibleRecaptcha/Extension/InvisibleRecaptchaPluginTest.php [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+    /**
+      * @testdox  can check answer with an empty private key
+      *
+      * @return  void
+      *
+      * @since   __DEPLOY_VERSION__
+      */
+    public function testResponseEmptyPrivateKey()
+    {
+        $language   = $this->createStub(Language::class);
```
Comment (richard67) [inline tests/Unit/Plugin/Captcha/InvisibleRecaptcha/Extension/InvisibleRecaptchaPluginTest.php]:
```suggestion
        $language = $this->createStub(Language::class);
```

File: tests/Unit/Plugin/Captcha/InvisibleRecaptcha/Extension/InvisibleRecaptchaPluginTest.php [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+    /**
+      * @testdox  can detect spam
+      *
+      * @return  void
+      *
+      * @since   __DEPLOY_VERSION__
+      */
+    public function testResponseSpam()
+    {
+        $language   = $this->createStub(Language::class);
```
Comment (richard67) [inline tests/Unit/Plugin/Captcha/InvisibleRecaptcha/Extension/InvisibleRecaptchaPluginTest.php]:
```suggestion
        $language = $this->createStub(Language::class);
```

File: tests/Unit/Plugin/Captcha/InvisibleRecaptcha/Extension/InvisibleRecaptchaPluginTest.php [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+    /**
+      * @testdox  can check successful answer
+      *
+      * @return  void
+      *
+      * @since   __DEPLOY_VERSION__
+      */
+    public function testFailedResponse()
+    {
+        $language   = $this->createStub(Language::class);
```
Comment (richard67) [inline tests/Unit/Plugin/Captcha/InvisibleRecaptcha/Extension/InvisibleRecaptchaPluginTest.php]:
```suggestion
        $language = $this->createStub(Language::class);
```

File: tests/Unit/Plugin/Captcha/InvisibleRecaptcha/Extension/InvisibleRecaptchaPluginTest.php [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+    /**
+      * @testdox  can return admin capabilities
+      *
+      * @return  void
+      *
+      * @since   __DEPLOY_VERSION__
+      */
+    public function testPrivacy()
+    {
+        $language   = $this->createStub(Language::class);
```
Comment (richard67) [inline tests/Unit/Plugin/Captcha/InvisibleRecaptcha/Extension/InvisibleRecaptchaPluginTest.php]:
```suggestion
        $language = $this->createStub(Language::class);
```

File: tests/Unit/Plugin/Captcha/Recaptcha/Extension/RecaptchaPluginTest.php [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+      * @testdox  can init the captcha
+      *
+      * @return  void
+      *
+      * @since   __DEPLOY_VERSION__
+      */
+    public function testInit()
+    {
+        $document   = new HtmlDocument();
+        $language   = $this->createStub(Language::class);
```
Comment (richard67) [inline tests/Unit/Plugin/Captcha/Recaptcha/Extension/RecaptchaPluginTest.php]:
```suggestion
        $document = new HtmlDocument();
        $language = $this->createStub(Language::class);
```

File: tests/Unit/Plugin/Captcha/Recaptcha/Extension/RecaptchaPluginTest.php [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+    /**
+      * @testdox  can init the captcha with an empty public key
+      *
+      * @return  void
+      *
+      * @since   __DEPLOY_VERSION__
+      */
+    public function testInitEmptyPublicKey()
+    {
+        $language   = $this->createStub(Language::class);
```
Comment (richard67) [inline tests/Unit/Plugin/Captcha/Recaptcha/Extension/RecaptchaPluginTest.php]:
```suggestion
        $language = $this->createStub(Language::class);
```

File: tests/Unit/Plugin/Captcha/Recaptcha/Extension/RecaptchaPluginTest.php [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+    /**
+      * @testdox  can check successful answer
+      *
+      * @return  void
+      *
+      * @since   __DEPLOY_VERSION__
+      */
+    public function testResponse()
+    {
+        $language   = $this->createStub(Language::class);
```
Comment (richard67) [inline tests/Unit/Plugin/Captcha/Recaptcha/Extension/RecaptchaPluginTest.php]:
```suggestion
        $language = $this->createStub(Language::class);
```

File: tests/Unit/Plugin/Captcha/Recaptcha/Extension/RecaptchaPluginTest.php [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+    /**
+      * @testdox  can check successful answer
+      *
+      * @return  void
+      *
+      * @since   __DEPLOY_VERSION__
+      */
+    public function testResponseWithCode()
+    {
+        $language   = $this->createStub(Language::class);
```
Comment (richard67) [inline tests/Unit/Plugin/Captcha/Recaptcha/Extension/RecaptchaPluginTest.php]:
```suggestion
        $language = $this->createStub(Language::class);
```

File: tests/Unit/Plugin/Captcha/Recaptcha/Extension/RecaptchaPluginTest.php [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+    /**
+      * @testdox  can check answer with an empty private key
+      *
+      * @return  void
+      *
+      * @since   __DEPLOY_VERSION__
+      */
+    public function testResponseEmptyPrivateKey()
+    {
+        $language   = $this->createStub(Language::class);
```
Comment (richard67) [inline tests/Unit/Plugin/Captcha/Recaptcha/Extension/RecaptchaPluginTest.php]:
```suggestion
        $language = $this->createStub(Language::class);
```

File: tests/Unit/Plugin/Captcha/Recaptcha/Extension/RecaptchaPluginTest.php [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+    /**
+      * @testdox  can detect spam
+      *
+      * @return  void
+      *
+      * @since   __DEPLOY_VERSION__
+      */
+    public function testResponseSpam()
+    {
+        $language   = $this->createStub(Language::class);
```
Comment (richard67) [inline tests/Unit/Plugin/Captcha/Recaptcha/Extension/RecaptchaPluginTest.php]:
```suggestion
        $language = $this->createStub(Language::class);
```

File: tests/Unit/Plugin/Captcha/Recaptcha/Extension/RecaptchaPluginTest.php [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+    /**
+      * @testdox  can check successful answer
+      *
+      * @return  void
+      *
+      * @since   __DEPLOY_VERSION__
+      */
+    public function testFailedResponse()
+    {
+        $language   = $this->createStub(Language::class);
```
Comment (richard67) [inline tests/Unit/Plugin/Captcha/Recaptcha/Extension/RecaptchaPluginTest.php]:
```suggestion
        $language = $this->createStub(Language::class);
```

File: tests/Unit/Plugin/Captcha/Recaptcha/Extension/RecaptchaPluginTest.php [hunk 17]
```diff
@@ -0,0 +1,290 @@
...
+    /**
+      * @testdox  can return admin capabilities
+      *
+      * @return  void
+      *
+      * @since   __DEPLOY_VERSION__
+      */
+    public function testPrivacy()
+    {
+        $language   = $this->createStub(Language::class);
```
Comment (richard67) [inline tests/Unit/Plugin/Captcha/Recaptcha/Extension/RecaptchaPluginTest.php]:
```suggestion
        $language = $this->createStub(Language::class);
```

**Comment by richard67** [comment]
RTC this time with the issue tracker.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39729">issues.joomla.org/tracker/joomla-cms/39729</a>.</sub>

**Comment by obuisard** [comment]
Thank you Allon @laoneo

**Pull request merged**